---
title: "outbreaker Summary Report"
output: html_document
date: "Generated:  `r format(Sys.time(), '%B %d, %Y, %H:%M')`<div style=\"height: 40px;\"></div>"
params:
  focal_list:
    input: file
    value: ""
  background_list:
    input: file
    value: ""
  snp_dists: 
    input: file
    value: ""
  snp_tree:
    input: file
    value: ""
---

```{r, include = F, message = F, warning = F, echo=F}
library(knitr)
library(tools)
library(Biostrings)
library(dplyr)
library(TraMineR)
library(DT)
library(ggtree)
library(ape)
library(ggplot2)
library(tidyverse)
library(magrittr)
library(heatmaply)
```


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Summary Statistics

```{r, echo = F, warning=F, message=F}
fasta_extensions <- c("fa", "fasta", "FA", "FASTA")
text_extensions <- c("txt", "TXT")

if (file_ext(params$focal_list) %in% fasta_extensions) {

  focal_read <- readDNAStringSet(params$focal_list)

  focal.fasta.names <- names(focal_read) %>%
  as.data.frame() %>%
  select("Focal Sequences" = ".")
  
  focal_input <- as.data.frame(focal.fasta.names)

  # datatable(as.data.frame(focal.fasta.names))
  
} else if (file_ext(params$focal_list) %in% text_extensions){
  focal_read <- read.csv(params$focal_list, header = FALSE,
                         na.strings=c("","NA"),
                         stringsAsFactors=FALSE,
                         sep="\t")

  focal_input <- as.data.frame(focal_read)
  colnames(focal_input) <- c("Sequence")
  # datatable(as.data.frame(focal_read))
} else {
  print("Error: focal sequences were not passed as either a txt file
        or a multi-FASTA")
  focal_input <- data.frame(Sequence = as.character())
}
```
<br>

```{r, echo = F, warning=F, message=F}

if (file_ext(params$background_list) %in% fasta_extensions) {

  background_read <- readDNAStringSet(file.path(params$background_list))

  background.fasta.names <- names(background_read) %>%
  as.data.frame() %>%
  select("Background Sequences" = ".")

  background_input <- as.data.frame(background.fasta.names)
  # datatable(as.data.frame(background.fasta.names))
} else if (file_ext(params$background_list) %in% text_extensions) {
  background_read <- read.csv(file.path(params$background_list), header = FALSE,
                         na.strings=c("","NA"),
                         stringsAsFactors=FALSE,
                         sep="\t")
  
  background_input <- as.data.frame(background_read)
  colnames(background_input) <- c("Sequence")
  # datatable(as.data.frame(background.fasta.names))
} else {
  background_input <- data.frame(Sequence = as.character())
}

```

```{r, echo = F, warning = F, message = F}
tree <- read.tree(params$snp_tree)

tree <- drop.tip(tree, c("MN908947"))

## Create a dataframe of the tree
tr.df <- fortify(tree)
## Create a list of tree labels
tr.df.labs <- tr.df %>%
  filter(isTip == "TRUE") %>%
  select(label)

if (file_ext(params$focal_list) %in% fasta_extensions) {
  tr.df.labs$category <- ifelse(tr.df.labs$label %in%
                                  as.vector(names(focal_read)), "Focal_Sequence", "Background_Sequence")
} else {
  tr.df.labs$category <- ifelse(tr.df.labs$label %in%
                                  focal_input$Sequence, "Focal_Sequence", "Background_Sequence")
}


```


Number of focal sequences input: **`r format (nrow(focal_input))`**
\
Number of background sequences input: **`r format (nrow(background_input))`**
\
\
Number of focal sequences retained: **`r format (nrow(subset(tr.df.labs, category == "Focal_Sequence")))`**
\
Number of background sequences retained: **`r format (nrow(subset(tr.df.labs, category == "Background_Sequence")))`**



## SNP Distances

```{r, echo = F, warning=F, message=F}
distances <- read.csv(params$snp_dists, header = FALSE,
                         na.strings=c("","NA"),
                         stringsAsFactors=FALSE,
                         sep=",") %>% filter(! grepl("MN908947", V1) &
                                               ! grepl("MN908947", V2))

filtered_w_background <- subset(distances, V1 %in% subset(tr.df.labs, category == "Focal_Sequence")$label &
                     ! V2 %in% subset(tr.df.labs, category == "Focal_Sequence")$label)

filtered_only_focal <- subset(distances, V1 %in% subset(tr.df.labs, category == "Focal_Sequence")$label &
                     V2 %in% subset(tr.df.labs, category == "Focal_Sequence")$label)

distance_frame_only_focal <- as.data.frame(as.data.frame(table(filtered_only_focal$V3)) %>% sapply( ., as.numeric ))
colnames(distance_frame_only_focal) <- c("SNP_Distance", "Frequency")

distance_frame_w_background <- as.data.frame(as.data.frame(table(filtered_w_background$V3)) %>% sapply( ., as.numeric ))

if (nrow(distance_frame_w_background) != 0) {
  colnames(distance_frame_w_background) <- c("SNP_Distance", "Frequency")
}

matrix <- as.matrix(spread(data = distances, key = V2, value = V3) %>% set_rownames(.$V1) %>% select(-V1) %>% replace(is.na(.), 0))

if (nrow(distance_frame_w_background) == 0 | any(is.infinite(distance_frame_w_background[,1]))) {
  closest_dist_w_background <- "NA (no background sequences)"
  farthest_dist_w_background <- "NA (no background sequences)"
} else {
  closest_dist_w_background <- as.numeric(min(distance_frame_w_background$SNP_Distance))
  farthest_dist_w_background <- as.numeric(max(distance_frame_w_background$SNP_Distance))
}

closest_dist_only_focal <- as.numeric(min(distance_frame_only_focal$SNP_Distance))

farthest_dist_only_focal <- as.numeric(max(distance_frame_only_focal$SNP_Distance))

```
<br>

Closest distance apart between two or more focal sequences: **`r format (closest_dist_only_focal)`**
\
Furthest distance apart between two or more focal sequences **`r format (farthest_dist_only_focal)`**


Closest distance apart between one more focal and one or more background sequences: **`r format (closest_dist_w_background)`**
\
Furthest distance apart between one more focal and one or more background sequences: **`r format (farthest_dist_w_background)`**

<br>

```{r, echo = F, warning=F, message=F, fig.width = 12, fig.height = 10}
if (nrow(matrix() >= 0)) {
  heatmaply(matrix, fontsize_row = 0, fontsize_col = 0, showticklabels = F)
} else {
  print("There was an error in loading the SNP distance heatmap")
}
```

<br>


## Phylogenetics

Focal sequences are coloured in red, while background sequences are coloured in black. 

```{r echo=F, fig.height=11, fig.width=18, message=FALSE, warning=FALSE}

annotated_tree <- ggtree(tree, size = 0.5) %<+% tr.df.labs +
  geom_tiplab(aes(x=x, subset=category == "Focal_Sequence", label = label),  size = 3, offset = 0, colour = "red") +
  geom_tiplab(aes(x=x, subset=category == "Background_Sequence", label = label),  size = 3, offset = 0, colour = "black") +
  theme(plot.margin = unit(c(0,0,0,0), "cm"),
        legend.text=element_text(size=10),
        legend.position = c(0.9, 0.9),
        legend.title=element_text(size=8, face = "bold"),
        legend.box = "vertical",
        legend.box.margin = margin(0.0,0.0,0.0,0.0,"cm"),
        legend.box.background = element_rect(colour = "grey50"),
        legend.spacing.y = unit(0.1, "cm"),
        legend.key.width = unit(0.3, "cm"),
        legend.key.height = unit(0.3, "cm"),
        legend.spacing.x = unit(0.3, 'cm'),
        legend.key.size = unit(3,"lines")) +
  guides(color = guide_legend(override.aes = list(size = 1.75)))

annotated_tree

```


