---
title: "outbreaker Summary Report"
author: "BCC PHO"
output: html_document
date: "Updated:  `r format(Sys.time(), '%B %d, %Y')`<div style=\"height: 40px;\"></div>"
params:
  focal_list:
    input: file
    value: ""
  background_list:
    input: file
    value: ""
  snp_dists: 
    input: file
    value: ""
  snp_tree:
    input: file
    value: ""
---

```{r, include = F, message = F, warning = F, echo=F}
library(knitr)
library(tools)
library(Biostrings)
library(dplyr)
library(TraMineR)
library(DT)
library(ggtree)
library(ape)
library(ggplot2)
library(tidyverse)
library(magrittr)
library(heatmaply)
```


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Focal Sequences

```{r, echo = F, warning=F, message=F}
fasta_extensions <- c("fa", "fasta", "FA", "FASTA")

if (file_ext(params$focal_list) %in% fasta_extensions) {

  focal_read <- readDNAStringSet(params$focal_list)

  focal.fasta.names <- names(focal_read) %>%
  as.data.frame() %>%
  select("Focal Sequences" = ".")

  datatable(as.data.frame(focal.fasta.names))
  
} else {
  focal_read <- read.csv(params$focal_list, header = FALSE,
                         na.strings=c("","NA"),
                         stringsAsFactors=FALSE,
                         sep="\t")

  datatable(as.data.frame(focal.fasta.names))
}
```
## Background Sequences

```{r, echo = F, warning=F, message=F}

if (file_ext(params$background_list) %in% fasta_extensions) {

  background_read <- readDNAStringSet(file.path(params$focal_list))

  background.fasta.names <- names(background_read) %>%
  as.data.frame() %>%
  select("Background Sequences" = ".")

  datatable(as.data.frame(background.fasta.names))
} else {
  background_read <- read.csv(file.path(params$background_list), header = FALSE,
                         na.strings=c("","NA"),
                         stringsAsFactors=FALSE,
                         sep="\t")

   datatable(as.data.frame(background.fasta.names))
}
```

## SNP Distances

```{r, echo = F, warning=F, message=F, fig.width = 12, fig.height = 10}
distances <- read.csv(params$snp_dists, header = FALSE,
                         na.strings=c("","NA"),
                         stringsAsFactors=FALSE,
                         sep=",")

distance_frame <- as.data.frame(table(distances$V3))
colnames(distance_frame) <- c("SNP Distance", "Frequency")

matrix <- as.matrix(spread(data = distances, key = V2, value = V3) %>% set_rownames(.$V1) %>% select(-V1) %>% replace(is.na(.), 0))

heatmaply(matrix, fontsize_row = 0, fontsize_col = 0, showticklabels = F)

```

## Phylogenetics

```{r echo=F, fig.height=11, fig.width=18, message=FALSE, warning=FALSE}

tree <- read.tree(params$snp_tree)

tree <- drop.tip(tree, c("MN908947"))

## Create a dataframe of the tree
tr.df <- fortify(tree)
## Create a list of tree labels
tr.df.labs <- tr.df %>%
  filter(isTip == "TRUE") %>%
  select(label)
  
tr.df.labs$category <- ifelse(tr.df.labs$label %in% as.vector(names(focal_read)), "Focal_Sequence", "Background_Sequence")

annotated_tree <- ggtree(tree, size = 0.25) %<+% tr.df.labs +
  geom_tiplab(aes(x=x, subset=category == "Focal_Sequence", label = label),  size = 2, offset = 0, colour = "red") +
  geom_tiplab(aes(x=x, subset=category == "Background_Sequence", label = label),  size = 2, offset = 0, colour = "black") +
  theme(plot.margin = unit(c(0,0,0,0), "cm"),
        legend.text=element_text(size=10),
        legend.position = c(0.9, 0.9),
        legend.title=element_text(size=8, face = "bold"),
        legend.box = "vertical",
        legend.box.margin = margin(0.0,0.0,0.0,0.0,"cm"),
        legend.box.background = element_rect(colour = "grey50"),
        legend.spacing.y = unit(0.1, "cm"),
        legend.key.width = unit(0.3, "cm"),
        legend.key.height = unit(0.3, "cm"),
        legend.spacing.x = unit(0.3, 'cm'),
        legend.key.size = unit(3,"lines")) +
  guides(color = guide_legend(override.aes = list(size = 1.75)))

annotated_tree

```


